src/elements/element.rectangle.js

```diff
@@ -54,95 +56,183 @@ function getBarBounds(bar) {
+	var t, r, b, l;
 
-		if (!vm.horizontal) {
-			// bar
-			left = vm.x - vm.width / 2;
-			right = vm.x + vm.width / 2;
-			top = vm.y;
-			bottom = vm.base;
-			signX = 1;
-			signY = bottom > top ? 1 : -1;
-			borderSkipped = vm.borderSkipped || 'bottom';
-		} else {
-			// horizontal bar
-			left = vm.base;
-			right = vm.x;
-			top = vm.y - vm.height / 2;
-			bottom = vm.y + vm.height / 2;
-			signX = right > left ? 1 : -1;
-			signY = 1;
-			borderSkipped = vm.borderSkipped || 'left';
+	if (helpers.isObject(value)) {
+		t = +value.top || 0;
+		r = +value.right || 0;
+		b = +value.bottom || 0;
+		l = +value.left || 0;
+	} else {
+		t = r = b = l = +value || 0;
+	}
+
+	return {
+		top: Math.min(maxHeight, skipped === 'top' ? 0 : t),
+		right: Math.min(maxWidth, skipped === 'right' ? 0 : r),
+		bottom: Math.min(maxHeight, skipped === 'bottom' ? 0 : b),
+		left: Math.min(maxWidth, skipped === 'left' ? 0 : l)
+	};
+}
+
+function flip(orig, v1, v2) {
+	return orig === v1 ? v2 : orig === v2 ? v1 : orig;
+}
+
+function parseBorderSkipped(bar) {
+	var vm = bar._view;
+	var vertical = isVertical(bar);
+	var borderSkipped = valueOrDefault(vm.borderSkipped, vertical ? 'bottom' : 'left');
+
+	if (vertical && vm.base < vm.y) {
+		borderSkipped = flip(borderSkipped, 'bottom', 'top');
+	}
+	if (!vertical && vm.base > vm.x) {
+		borderSkipped = flip(borderSkipped, 'left', 'right');
+	}
+	return borderSkipped;
+}
+
+function buildBorderSections(bounds, inner, border, radius) {
+	var top = 'top';
+	var left = 'left';
+	var right = 'right';
+	var bottom = 'bottom';
+	var borders = [top, right, bottom, left];
+	var corners = [[left, top], [right, top], [right, bottom], [left, bottom]];
+	var sections = [];
+	var current = [];
+	var count = 0;
+	var startIdx = 0;
+	var i, b, n, p, c1, c2;
+
+	radius = radius || 0;
+
+	for (i = 0; i < 4; ++i) {
+		if (!border[borders[i]]) {
+			startIdx = i;
+			break;
 		}
+	}
 
-		// Canvas doesn't allow us to stroke inside the width so we can
-		// adjust the sizes to fit if we're setting a stroke on the line
-		if (borderWidth) {
-			// borderWidth shold be less than bar width and bar height.
-			var barSize = Math.min(Math.abs(left - right), Math.abs(top - bottom));
-			borderWidth = borderWidth > barSize ? barSize : borderWidth;
-			var halfStroke = borderWidth / 2;
-			// Adjust borderWidth when bar top position is near vm.base(zero).
-			var borderLeft = left + (borderSkipped !== 'left' ? halfStroke * signX : 0);
-			var borderRight = right + (borderSkipped !== 'right' ? -halfStroke * signX : 0);
-			var borderTop = top + (borderSkipped !== 'top' ? halfStroke * signY : 0);
-			var borderBottom = bottom + (borderSkipped !== 'bottom' ? -halfStroke * signY : 0);
-			// not become a vertical line?
-			if (borderLeft !== borderRight) {
-				top = borderTop;
-				bottom = borderBottom;
+	for (i = 0; i < 4; ++i) {
+		b = borders[(i + startIdx) % 4];
+		if (border[b]) {
+			n = borders[(i + startIdx + 1) % 4];
```

Comment:
Longer variable names or some sort of explanation would help. It's not immediately clear to me what `n` is. Same with many of the others

Reply:
done